cc_corpus/index.py

```diff
@@ -0,0 +1,200 @@
+    a particular file.
+    """
+    file_name: str
+    offset: int
+    length: int
+
+
+def read_cluster_idx(cluster_idx: Path) -> Generator[Cluster]:
+    """Enumerates the clusters (lines) in a cluster.idx file."""
+    with openall(cluster_idx, 'rt') as inf:
+        for line in map(str.strip, inf):
+            yield Cluster.from_line(line)
+
+
+def find_tld_in_index(tld: str, cluster_idx: Path) -> list[Cluster]:
+    """
+    Finds all clusters that end in a particular TLD in cluster.idx.
+
+    .. note::
+
+        The cluster that comes before the first one with _tld_ is included in
+        the list. This is because the URL in the index file is the first in
+        a block of 3,000, so it is possible that the previous cluster already
+        has URLs with the TLD in question.
+    """
+    last_cluster = None
+    clusters = []
+    for cluster in read_cluster_idx(cluster_idx):
+        ctld = cluster.surt[:cluster.surt.find(',')]
+        if ctld < tld:
+            last_cluster = cluster
+        elif ctld >= tld:
+            if last_cluster is not None:
+                clusters.append(last_cluster)
+                last_cluster = None
+            if ctld == tld:
+                clusters.append(cluster)
+            else:
+                break
+    return clusters
+
+
+def ranges_from_clusters(
+    clusters: list[Cluster], max_clusters: int = 0
+) -> Generator[list[FileRange]]:
+    """
+    Creates :class:`FileRange`s from the clusters acquired via
+    :func:`find_tld_in_index`.
+
+    :param clusters: a list of clusters, ordered by SURT and offset.
+    :param max_clusters: the maximum number of clusters in a batch. The
+                         default is ``0``, meaning no limit (i.e. all
+                         clusters are returned in a single batch).
+    :return: yields lists of :class:`FileRange`s.
+    """
+    def range_from_clusters(file_name: str, file_clusters: Iterable[Cluster]):
+        """
+        Returns a single range for an :class:`Iterable` of :class:`Cluster`s.
+        """
+        start, end = None, None
+        for cluster in file_clusters:
+            if start is None:
+                start = cluster.offset
+                end = start + cluster.length
+            else:
+                if cluster.offset != end:
+                    raise ValueError(f'Discontinuous cluster {cluster.surt}: '
+                                     f'{cluster.offset=} instead of {end=}!')
+                else:
+                    end += cluster.length
+        return FileRange(file_name, start, end - start)
+
+    for file_name, file_clusters in groupby(clusters,
+                                            key=attrgetter('file_name')):
+        if max_clusters > 0:
+            for batch in batched(file_clusters, max_clusters):
+                yield range_from_clusters(file_name, batch)
+        else:
+            yield range_from_clusters(file_name, file_clusters)
+
+
+class BatchWriter:
```

Comment:
TBH, now that you wrote the third BatchWriter class, with overlapping functionallity, we should create a parent class, that has the shared fields and functions, and use class inheritance.

Because right now we are repeating a lot of code in these BatchWriter classes.

Reply:
Yeah, definitely. In another PR. :)